internal/services/oidc.go

```diff
@@ -0,0 +1,206 @@
+	}
+	idToken, err := osvc.validateIDToken(ctx, oauth2Token)
+	if err != nil {
+		return models.Tokens{}, err
+	}
+
+	idTokenClaims := new(json.RawMessage)
+	if err := idToken.Claims(&idTokenClaims); err != nil {
+		return models.Tokens{}, errors.New("An error occurred while validating ID Token. Error: " + err.Error())
+	}
+
+	userInfo, err := osvc.getUserInfo(ctx, oauth2Token)
+	if err != nil {
+		return models.Tokens{}, errors.New("Failed to get UserInfo. Error: " + err.Error())
+	}
+
+	cc := models.CustomClaims{}
+	if err := idToken.Claims(&cc); err != nil {
+		return models.Tokens{}, errors.New("An unexpected error has occurred. Error: " + err.Error())
+	}
+
+	return models.Tokens{
+		OAuth2Token:   oauth2Token,
+		IDTokenClaims: idTokenClaims,
+		UserInfo:      userInfo,
+		CustomClaims:  cc,
+	}, nil
+}
+
+// IsFlowSecure ensures the flow is secure and then, returns a JWT token.
+func (osvc *OIDCService) IsFlowSecure(state string, token models.Tokens) (bool, error) {
+	if state == "" {
+		return false, errors.New("unable to follow a secure flow without the state returned from the OIDC Provider")
+	}
+
+	session, err := osvc.sessionStorage.GetSession(state)
+	if err != nil {
+		return false, err
+	}
+
+	nonceToken, err := osvc.getValueFromToken("nonce", token)
+	if err != nil {
+		return false, err
+	}
+	if !session.MatchNonce(nonceToken.(string)) {
+		return false, errors.New("nonce did not match")
+	}
+
+	return true, nil
+}
+
+// CreateJSONData creates json data to display as body content.
+func (osvc *OIDCService) CreateJSONData(token models.Tokens) ([]byte, error) {
+	jwt, err := osvc.createJWT(&token.CustomClaims)
+	if err != nil {
+		return nil, errors.New("an error occurred while creating a JWT. Error: " + err.Error())
+	}
+
+	// If all goes well, then create the JSON data to display as body content.
+	jsonBody, err := json.Marshal(jwt)
+	if err != nil {
+		return nil, errors.New("an error occurred while validating some tokens. Error: " + err.Error())
+	}
+
+	return jsonBody, nil
+}
+
+// getUserInfo gets User Info from the OAuth2 Token.
+func (osvc *OIDCService) getUserInfo(ctx context.Context, oauth2Token *oauth2.Token) (*oidc.UserInfo, error) {
+	userInfo, err := osvc.client.Provider.UserInfo(ctx, oauth2.StaticTokenSource(oauth2Token))
+	if err != nil {
+		return nil, err
+	}
+
+	return userInfo, nil
+}
+
+// getValueFromToken gets the nonce from the ID Token.
+func (osvc *OIDCService) getValueFromToken(value string, t models.Tokens) (interface{}, error) {
+	var m map[string]interface{}
+	if err := json.Unmarshal(*t.IDTokenClaims, &m); err != nil {
+		return "", errors.New("an unexpected error occurred when getting the nonce from ID Token")
+	}
+
+	if key, ok := m[value]; ok {
+		return key, nil
+	}
+	return "", nil
+}
+
+// validateIDToken validates the ID token.
+func (osvc *OIDCService) validateIDToken(ctx context.Context, oauth2Token *oauth2.Token) (*oidc.IDToken, error) {
```

Comment:
do we need to use pointers here?

Reply:
`osvc.client.Verifier.Verify` returns a `*oidc.IDToken`.
 I think we need or I can return `*idToken`. 🤔